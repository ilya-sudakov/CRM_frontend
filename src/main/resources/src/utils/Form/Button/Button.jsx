import React, { useEffect, useState } from 'react';
import './Button.scss';

const Button = (props) => {
    const [className, setClassName] = useState('button');

    useEffect(() => {
        if (props.inverted) {
            setClassName(className + " button--inverted")
        }
    }, [])

    return (
        <button className={className + (props.className ? (" " + props.className) : '')} onClick={(event) => {
            event.preventDefault();
            // console.log(event);
            if (!props.isLoading) {
                props.onClick();
            }
        }}>
            {props.imgSrc && <img className={props.isLoading ? "button__img button__img--hidden" : "button__img"} src={props.imgSrc} alt="" />}
            <span className={props.isLoading ? "button__text button__text--hidden" : "button__text"}>{props.text ?? 'Нажмите'}</span>
            {props.isLoading && <div class="button__circle"></div>}
        </button>
    );
};

export default Button;