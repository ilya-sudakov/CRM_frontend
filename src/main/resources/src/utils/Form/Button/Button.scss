@import "../../../variables.scss";

.button {
  background-color: $button__background--inverted;
  color: $secondary-text-color;
  border-radius: 5px;
  padding: 10px 15px;
  cursor: pointer;
  font-size: 90%;
  margin-right: 10px;
  transition: 0.2s ease-in-out;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  border: 1px solid transparent;
  text-align: center;
  outline: none;
  -moz-user-select: none;
  -khtml-user-select: none;
  -webkit-user-select: none;
  user-select: none;
  position: relative;

  &:hover {
    background-color: lighten($button__background--inverted, 5%);
    box-shadow: 0 4px 4px rgba(0, 0, 0, 0.22), 0 0px 5px rgba(0, 0, 0, 0.24);
  }

  &--inverted {
    background-color: $secondary-text-color;
    color: $primary-text-color;
    border: 1px solid $button__background--inverted;

    &__img {
      max-width: 20px;
      margin-right: 5px;
      filter: brightness(1) invert(0);
    }

    &:hover {
      color: $secondary-text-color;
    }

    .button__circle {
      border: 4px rgba(100, 100, 100, 0.5) solid !important;
      border-top: 4px #666666 solid !important;
    }
  }

  &__text {
    &--hidden {
      visibility: hidden;
    }
  }

  &__img {
    max-width: 20px;
    margin-right: 5px;
    filter: brightness(3) invert(1);
    &--hidden {
      visibility: hidden;
    }
  }

  .button__circle {
    display: block;
    position: absolute;
    top: calc(50% - (22px / 2));
    left: calc(50% - (20px / 2));
    height: 15px;
    width: 15px;
    // margin: -25px 0 0 -25px;
    border: 4px rgba(255, 255, 255, 0.5) solid;
    border-top: 4px #eeeeee solid;
    border-radius: 999px;
    -webkit-animation: spin2 1s infinite linear;
    animation: spin2 1s infinite linear;
  }

  @-webkit-keyframes spin2 {
    from {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(359deg);
      transform: rotate(359deg);
    }
  }
  @keyframes spin2 {
    from {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(359deg);
      transform: rotate(359deg);
      -webkit-transform: rotate(359deg);
      transform: rotate(359deg);
    }
  }
}
