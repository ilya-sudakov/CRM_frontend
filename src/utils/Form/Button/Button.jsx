import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './Button.scss';

const Button = ({
  inverted = false,
  className = '',
  isLoading = false,
  imgSrc = '',
  type = 'button',
  text = 'Нажмите',
  onClick = null,
  style = {},
  isRecent = false,
}) => {
  const [newClassName] = useState(
    `button ${inverted ? ` button--inverted` : ''}`,
  );

  const handleClick = (event) => {
    event.preventDefault();
    if (!isLoading) {
      return onClick();
    }
  };

  return (
    <button
      className={`${newClassName} ${className ?? ''} ${
        isRecent ? 'main-window__button--recent' : ''
      }`}
      onClick={handleClick}
      type={type ? type : 'button'}
      style={{ ...style }}
    >
      {imgSrc && (
        <img
          className={`button__img ${isLoading ? 'button__img--hidden' : ''}`}
          src={imgSrc}
          alt=""
        />
      )}
      <span
        className={`button__text ${isLoading ? 'button__text--hidden' : ''}`}
      >
        {text}
      </span>
      {isLoading && <div className="button__circle"></div>}
    </button>
  );
};

export default Button;

Button.propTypes = {
  inverted: PropTypes.bool,
  className: PropTypes.string,
  isLoading: PropTypes.bool,
  isRecent: PropTypes.bool,
  imgSrc: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  type: PropTypes.string,
  text: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func,
};
