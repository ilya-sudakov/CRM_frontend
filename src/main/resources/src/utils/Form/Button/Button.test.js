import React from 'react'
import Button from './Button.jsx'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'

describe('Button component', () => {
  afterEach(cleanup)

  it('matches snapshot', () => {
    const { asFragment } = render(<Button />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders w/o props', () => {
    render(<Button />)
  })

  it('renders button w/ inverted styles', () => {
    render(<Button inverted />)
    const button = screen.getByRole('button')
    expect(button.classList.contains('button--inverted')).toBe(true)
  })

  it('handles onClick event', () => {
    let temp = 0
    render(<Button onClick={() => temp++} />)
    fireEvent.click(screen.getByRole('button'))
    expect(temp === 1).toBe(true)
  })

  it('ignores onClick event if loading is happening', () => {
    let temp = 0
    render(<Button onClick={() => temp++} isLoading={true} />)
    fireEvent.click(screen.getByRole('button'))
    expect(temp === 0).toBe(true)
  })
})
