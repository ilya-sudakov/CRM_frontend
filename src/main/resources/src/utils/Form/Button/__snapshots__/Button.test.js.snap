// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="button  "
    type="button"
  >
    <span
      class="button__text "
    >
      Нажмите
    </span>
  </button>
</DocumentFragment>
`;
